godtools/App/Services/UserAuthentication/Okta/OktaUserAuthentication.swift

```diff
@@ -0,0 +1,161 @@
+//
+//  OktaUserAuthentication.swift
+//  godtools
+//
+//  Created by Levi Eggert on 12/1/21.
+//  Copyright © 2021 Cru. All rights reserved.
+//
+
+import UIKit
+import OktaAuthentication
+
+class OktaUserAuthentication: UserAuthenticationType {
```

Comment:
I wonder, why doesn't this live in the library? Nothing here seems GodTools specific. Is it because of where it lives in the flow of the app?

Reply:
Hey @wrandall22 it possibly could and I have been considering that.  The only conditions are 

1. OktaUserAuthentication assumes that Okta is configured with the offline_access in order for renewing authentication.  Although, I guess regardless if that scope was there are not this implementation would still function.

2. It assumes that auth user attributes are consistent between apps.  I'm not sure where those user attributes are defined and if they would match up between GodTools and MPDX.  That's a concern.

It would be nice to share this logic and that is something I'm considering.  It would also move the protocol UserAuthenticationType into the shared library which isn't that big of a deal.  I'd just want to make that more clear in the shared library.

Reply:
> OktaUserAuthentication assumes that Okta is configured with the offline_access in order for renewing authentication. Although, I guess regardless if that scope was there are not this implementation would still function.
I would think all mobile apps would want offline_access, right?

> It assumes that auth user attributes are consistent between apps. I'm not sure where those user attributes are defined and if they would match up between GodTools and MPDX. That's a concern.
What are some examples of these attributes? Maybe I could help you know where they are defined.

Reply:
Hey @wrandall22 yeah I think all apps would want the offline_access.
At the moment GodTools is using "email", "given_name", "grMasterPersonId", "family_name", "ssoguid".
MPDX is using: "email", "name", "grMasterPersonId", "ssoguid" which does match up with GodTools.  I just want to ensure they would stay consistent.

Reply:
I also would want to remove the ObservableValue and Signal dependencies from UserAuthenticationType and use closures instead.

Reply:
I see, so it sounds like those are attributes within Okta itself. Is `name` and `family_name`/`given_name` the same thing or are you using different fields for them? I suppose the easy way would be to include all the ones all the apps use and decide within each app whether or not to care about said field.

Reply:
Yeah those are defined in Okta.  I'm not sure about the name.  I need to run MPDX and see what fields it gives back to me.

Reply:
Another option besides including all the properties would be to have a class in the library with the common properties and use class extensions in the apps where necessary.

Reply:
Yeah that's another option which the shared library does provide through a CruOktaUser.swift as well as methods for fetching straight user json and even a generic Decodable type if clients want to have the shared library decode a type provided by the client.

It would be good to move the refresh authorization logic to the shared library and possibly the UserAuthenticationType protocol since both MPDX and GodTools support that.

Reply:
to add a little bit of Okta context to this discussion, For each app configured in Okta you can choose to map user profile attributes however you want. so `ssoguid` could be `sso_guid` for a different app. it's really down to conventions within the app config to get any consistency with profile attribute names.

Reply:
on android I did create a centralized definition of common profile attributes and accessors in gto-support: https://github.com/CruGlobal/android-gto-support/blob/master/gto-support-okta/src/main/kotlin/org/ccci/gto/android/common/okta/oidc/net/response/UserInfo.kt#L6-L19

Reply:
Our common spot for Java web is [here](https://github.com/CruGlobal/okta-client/blob/master/src/main/kotlin/org/cru/okta/client/model/user/OktaUser.kt#L6)